### PR TITLE
NIFI-7811: Prevent NPE and bad behavior when PrometheusReportingTask can't start

### PR DESCRIPTION
**Pull request by mattyb149**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

If PrometheusReportingTask can't start (for example, if there's a Jetty issue such as the port already being in use), it continues to be triggered which causes NullPointerExceptions to be logged with no cause. In the case that the PrometheusReportingTask cannot be started successfully, it should throw an exception during initialization. It will continue to attempt to initialize the reporting task (that's handled by the framework), but it will not call its onTrigger method. Some defensive code to prevent NPEs was added as well.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Review by ChrisSamo632 — COMMENTED** [review]
LGTM, thanks

File: nifi-nar-bundles/nifi-prometheus-bundle/nifi-prometheus-reporting-task/src/main/java/org/apache/nifi/reporting/prometheus/PrometheusRecordSink.java [hunk 1]
```diff
@@ -181,15 +183,23 @@ public WriteResult sendData(RecordSet recordSet, Map<String, String> attributes,
 
     @OnDisabled
     public void onStopped() throws Exception {
```
Comment (ottobackwards) [inline nifi-nar-bundles/nifi-prometheus-bundle/nifi-prometheus-reporting-task/src/main/java/org/apache/nifi/reporting/prometheus/PrometheusRecordSink.java:185]: this api would be a good place to use Optional<T>

**Comment by ottobackwards** [comment]
This looks good, I think using `Optional<T>` for the getPrometheusServer return and other places in here would be a cleaner approach.

+1 fwiw

**Review by pvillard31 — APPROVED** [review]

**Comment by pvillard31** [comment]
I was about to merge this one but I might have observed something we don't want when stopping NiFi. Trying to confirm. Please hold on before merging.

**Review by pvillard31 — CHANGES_REQUESTED** [review]
Not sure what we can do about it, but if starting two reporting tasks, I'll get the exception as expected for the second one. But then, if I stop the two reporting tasks and stop NiFi, then it seems there are some threads we can't stop properly - probably related to the Jetty server we tried to start but which reported "address bind already in use":

````
$ nifi stop

Java home: /Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home
NiFi home: /Users/pierrevillard/dev/github/nifi/nifi-assembly/target/nifi-1.13.0-SNAPSHOT-bin/nifi-1.13.0-SNAPSHOT

Bootstrap Config File: /Users/pierrevillard/dev/github/nifi/nifi-assembly/target/nifi-1.13.0-SNAPSHOT-bin/nifi-1.13.0-SNAPSHOT/conf/bootstrap.conf

2020-09-15 21:04:26,023 INFO [main] org.apache.nifi.bootstrap.Command Apache NiFi has accepted the Shutdown Command and is shutting down now
2020-09-15 21:04:26,365 INFO [main] org.apache.nifi.bootstrap.Command Waiting for Apache NiFi to finish shutting down...
2020-09-15 21:04:28,380 INFO [main] org.apache.nifi.bootstrap.Command Waiting for Apache NiFi to finish shutting down...
2020-09-15 21:04:30,391 INFO [main] org.apache.nifi.bootstrap.Command Waiting for Apache NiFi to finish shutting down...
2020-09-15 21:04:32,404 INFO [main] org.apache.nifi.bootstrap.Command Waiting for Apache NiFi to finish shutting down...
2020-09-15 21:04:34,420 INFO [main] org.apache.nifi.bootstrap.Command Waiting for Apache NiFi to finish shutting down...
2020-09-15 21:04:36,436 INFO [main] org.apache.nifi.bootstrap.Command Waiting for Apache NiFi to finish shutting down...
2020-09-15 21:04:38,451 INFO [main] org.apache.nifi.bootstrap.Command Waiting for Apache NiFi to finish shutting down...
2020-09-15 21:04:40,462 INFO [main] org.apache.nifi.bootstrap.Command Waiting for Apache NiFi to finish shutting down...
2020-09-15 21:04:42,474 INFO [main] org.apache.nifi.bootstrap.Command Waiting for Apache NiFi to finish shutting down...
2020-09-15 21:04:44,486 INFO [main] org.apache.nifi.bootstrap.Command Waiting for Apache NiFi to finish shutting down...
2020-09-15 21:04:46,500 INFO [main] org.apache.nifi.bootstrap.Command Waiting for Apache NiFi to finish shutting down...
2020-09-15 21:04:46,510 WARN [main] org.apache.nifi.bootstrap.Command NiFi has not finished shutting down after 20 seconds. Killing process.
2020-09-15 21:04:46,527 INFO [main] org.apache.nifi.bootstrap.Command NiFi has finished shutting down.
````

If starting and stopping just one reporting task, resources are stopped correctly:

````
$ nifi stop

Java home: /Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home
NiFi home: /Users/pierrevillard/dev/github/nifi/nifi-assembly/target/nifi-1.13.0-SNAPSHOT-bin/nifi-1.13.0-SNAPSHOT

Bootstrap Config File: /Users/pierrevillard/dev/github/nifi/nifi-assembly/target/nifi-1.13.0-SNAPSHOT-bin/nifi-1.13.0-SNAPSHOT/conf/bootstrap.conf

2020-09-15 21:08:26,348 INFO [main] org.apache.nifi.bootstrap.Command Apache NiFi has accepted the Shutdown Command and is shutting down now
2020-09-15 21:08:26,686 INFO [main] org.apache.nifi.bootstrap.Command Waiting for Apache NiFi to finish shutting down...
2020-09-15 21:08:28,699 INFO [main] org.apache.nifi.bootstrap.Command NiFi has finished shutting down.
````

**Review by pvillard31 — APPROVED** [review]

**Comment by pvillard31** [comment]
Merged to main, thanks @mattyb149. Probably worth being included in 1.12.1 @joewitt.

**Pull request closed without merging**